doc/adr/0016-end-to-end-testing-strategy.rst

```diff
@@ -0,0 +1,277 @@
+   When ready we will begin incorporating it as a replacement for ``cardano-api``.
+
+.. _`argument 6`:
+
+6. Although ``cardano-ledger-api`` and ``cardano-node-client`` are under test it isn't feasible
+   to expect thorough coverage of all ledger and node functionality, such as staking and update
+   proposals, because the primary focus is to satisfy end-to-end testing requirements for ``Plutus``.
+   Fortunately, much of that functionality is already being covered by ``cardano-node-tests``.
+
+.. _`argument 7`:
+
+7. Initially, a few tests will be created without depending on `Plutus-apps
+   <https://github.com/input-output-hk/Plutus-apps/>`_.
+   This entails building transactions with ``cardano-ledger-api`` and waiting for on-chain events
+   using ``cardano-node-client`` without use of the Contract api or the constraints library.
+   This approach allows us to build specific transactions, which is especially useful for testing
+   edge-cases and error scenarios that off-chain tooling may prohibit.
+   However, this approach will require more boilerplate code and this could negatively impact
+   readability of the tests.
+   Having assessed this approach, we may then decide to depend on ``Plutus-apps`` for the
+   Contract api, which would give a uniform interface for off-chain code such as different node
+   backends (private and public testnets, and emulator) and chain-indexer queries
+   (``cardano-node-client`` or Marconi in future).
+   It should also reduce the amount of boilerplate code and provide additional features such as
+   trace logging.
+
+.. _`argument 8`:
+
+8. There's value continuing to test ``cardano-cli`` with Plutus transactions for specific cli
+   flags and the cli's error handling with script evaluation.
+   Some examples of tests that should be covered:
+
+   * Cli flags that require use of Plutus scripts E.g. tx-out-reference-script-file or 
+     calculate-plutus-script-cost
+   * Cli behaviour when script evaluation passes. This could be displaying expected fee correctly.
+   * Cli behaviour when script evaluation fails. This can be how different types errors are formatted. 
+
+9. At some point, we may wish to incorporate the `cardano-node-emulator
+   <https://github.com/input-output-hk/Plutus-apps/pull/831>`_ as an alternative to ``cardano-testnet``.
+   This would enable us to run property based tests due to the node being much faster without consensus.
+   With CHaP, ``cardano-node-emulator`` would be released as a separate component, so no need to depend
+   on ``Plutus-apps``.
+
+10. We reserve the option of including additional packages to test from ``Plutus-apps`` at a later stage.
+
+Pros of building and maintaining our own test framework
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* Plutus tools team will have full ownership of the end-to-end test environment and its priorities
+  for ``Plutus``.
+* Plutus scripts can be defined alongside the tests. ``cardano-node-tests`` requires
```

Comment:
This is not necessarily an advantage. By compiling plutus scripts separately, we can publish them separately, and other frameworks (like CTL) could use them to test their frameworks.

Reply:
That's kinda what they do already but it could be better organised, There's also the need to produce json files for each typed script data argument, extra steps.